# Analyzing data sets from AppleStore and GooglePlay
* Our company builds only Android and iOS mobile apps that are free to download, which makes the main revenue stream comes from in-app ads. Thus, it is interesting to the company gathering as many users as possible. 
* This project is about to analyze data to help our developers **understand what type of apps are likely to attract more users on Google Play and App Store.**
* My main goal in this project is to gain experience with structural elements and functions of Python in real situations and problems.

In [3]:
from csv import reader
def open_data(file_name):
    open_file = open(file_name)
    readed_file = reader(open_file)
    data = list(readed_file)
    return data

app_data = open_data('AppleStore.csv')
google_data = open_data('googleplaystore.csv')

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [5]:
explore_data(dataset=google_data[1:], start=0, end=5, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13


In [6]:
explore_data(dataset=app_data[1:], start=0, end=5, rows_and_columns=True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


In [7]:
print(app_data[0])
print(google_data[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The columns that could help us with our analysis are:

**app_data**: size_bytes; price; rating_count_tot; user_rating; cont_rating; prime_genre

**google_data**: Category; Rating; Reviews; Size; Installs; Type; Content Rating; Genres

In [8]:
print(google_data[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The row above had a missing Category data, so we deleted it to clean the inaccuracies of our dataset and don't mess our data analysis. This error was found by the kaggle discussion section of this database. [Link] (https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015)

In [9]:
del google_data[10473]

In [10]:
explore_data(dataset=google_data, start=10470, end=10475)

['TownWiFi | Wi-Fi Everywhere', 'COMMUNICATION', '3.9', '2372', '58M', '500,000+', 'Free', '0', 'Everyone', 'Communication', 'August 2, 2018', '4.2.1', '4.2 and up']


['Jazz Wi-Fi', 'COMMUNICATION', '3.4', '49', '4.0M', '10,000+', 'Free', '0', 'Everyone', 'Communication', 'February 10, 2017', '0.1', '2.3 and up']


['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['Sat-Fi Voice', 'COMMUNICATION', '3.4', '37', '14M', '1,000+', 'Free', '0', 'Everyone', 'Communication', 'November 21, 2014', '2.2.1.5', '2.2 and up']




Besides this missing data innacuracy, the dataset has duplicate entries, which will also contaminate our analysis since we don't want count apps more than once, so we need to remove it. 

In [11]:
duplicate_data = []
unique_data = []

for row in google_data[1:]:
    appname = row[0]

    if appname in unique_data:
        duplicate_data.append(appname)
    else:
        unique_data.append(appname)

print('Number of duplicate apps: ', len(duplicate_data))        
print('\n')
print('Number of unique apps: ', len(unique_data))   
print('\n')   
print(duplicate_data[:15])
print('\n')
print(unique_data[:10])
    

Number of duplicate apps:  1181


Number of unique apps:  9659


['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun']


 In these cases, we have to determine which row of these duplicates we need to take.

In [12]:
for app in google_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Analysing the duplicates above we can see that the number of Reviews is the only column values changing. It is because these data was taken in different times, so the higher the number of Reviews the more recent this information was taken. This way, we gonna get only the most recent one (that's our filter).

So, let's remove these duplicates.

In [13]:
reviews_max = {}
for app in google_data[1:]:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


In [14]:
android_clean = []
already_added = []
for app in google_data[1:]:
    name = app[0]
    n_reviews = float(app[3])
    if (name not in already_added) and (n_reviews == reviews_max[name]):
        android_clean.append(app)
        already_added.append(name)
        
print(android_clean[:10])
google_data_clean = android_clean

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up'], ['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+', '

We needed to add this supplementary condition (already_added) to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [15]:
explore_data(dataset=android_clean, start=0, end=5, rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9659
Number of columns: 13


Now, let's also check if there are duplicates in app store dataset.

In [16]:
duplicate_data = []
unique_data = []

for row in app_data[1:]:
    appname = row[0]

    if appname in unique_data:
        duplicate_data.append(appname)
    else:
        unique_data.append(appname)

print('Number of duplicate apps: ', len(duplicate_data))        
print('\n')
print('Number of unique apps: ', len(unique_data))   
print('\n')   
print(duplicate_data[:15])
print('\n')
print(unique_data[:10])

Number of duplicate apps:  0


Number of unique apps:  7197


[]


['284882215', '389801252', '529479190', '420009108', '284035177', '429047995', '282935706', '553834731', '324684580', '343200656']


Thus, there are no duplicates in this data set, so we can proceed.

In our company, we develop apps toward an English-speaking audience.
So, we aren't interested in keeping apps, for instance, in japanese or chinese. We will remove them.

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

However, if we consider only that just one character ascii number is greater than 127, then we could be filtering english apps that have a special character in its name, such as 😜 or ™. Thus, to minimize this impact, we will filter apps with more than 3 character ascii number > 127.

In [17]:
print(ord('a'))
print(ord('9'))
print(ord('P'))

97
57
80


In [18]:
def english_only(string):
    non_English_counter = 0
    for character in string:
        ascii_number = ord(character)
        if ascii_number > 127:
            non_English_counter += 1
        if non_English_counter > 3:
            return False
    return True
    

In [19]:
print(english_only('Instagram'))
print(english_only('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_only('Docs To Go™ Free Office Suite'))
print(english_only('Instachat 😜'))

True
False
True
True


In [20]:
appdata_english_apps = []
appdata_non_english = []
for row in app_data[1:]:
    appname = str(row[1])
    is_english = english_only(appname)
    if is_english == True:
        appdata_english_apps.append(row)
    else:
        appdata_non_english.append(row)
        
print('Number of English Apps in Apple Store data: ', len(appdata_english_apps))
print('Number of Non-English Apps in Apple Store data: ', len(appdata_non_english))

Number of English Apps in Apple Store data:  6183
Number of Non-English Apps in Apple Store data:  1014


In [21]:
googledata_english_apps = []
googledata_non_english = []
for row in android_clean:
    appname = str(row[0])
    is_english = english_only(appname)
    if is_english == True:
        googledata_english_apps.append(row)
    else:
        googledata_non_english.append(row)

print('Number of English Apps in Google Play filtered data: ', len(googledata_english_apps))
print('Number of Non-English Apps in Google Play filtered data: ', len(googledata_non_english))

Number of English Apps in Google Play filtered data:  9614
Number of Non-English Apps in Google Play filtered data:  45


Now, to finish our data cleaning step, we'll isolate the free apps from these data sets. 

In [22]:
googledata_freeapps = []
googledata_nonfree = []
for row in googledata_english_apps:
    price = str(row[7])
    if price == '0':
        googledata_freeapps.append(row)
    else:
        googledata_nonfree.append(row)
        
print('Number of Free Apps in Google Play filtered data: ', len(googledata_freeapps))
print('Number of Non-Free Apps in Google Play filtered data: ',len(googledata_nonfree))

Number of Free Apps in Google Play filtered data:  8864
Number of Non-Free Apps in Google Play filtered data:  750


In [23]:
appdata_freeapps = []
appdata_nonfree = []
for row in appdata_english_apps:
    price = str(row[4])
    if price == '0.0':
        appdata_freeapps.append(row)
    else:
        appdata_nonfree.append(row)
        
print('Number of Free Apps in Apple Store filtered data: ', len(appdata_freeapps))
print('Number of Non-Free Apps in Apple Store filtered data: ',len(appdata_nonfree))

Number of Free Apps in Apple Store filtered data:  3222
Number of Non-Free Apps in Apple Store filtered data:  2961


# Analyzing the filtered data sets
So far, we cleaned our data sets in order to remove:
- Inaccurate data; 
- Duplicate app entries;
- non-English apps;
- non-Free apps.

Again, as we mentioned in the introduction, our company build only free apps and our main source of revenue consists of in-app ads.
To minimize risks and overhead, our validation strategy for an app idea is comprised of 3 steps:
- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Now, we are going to getting a sense of what are the most common genres for each market. We'll build frequency tables for a fey columns in our data sets.

**Google Play:** the columns **'Category'** and **'Genres'** seem to represent what the app is about, with the second one "tokenizing" some classifications for each apps.

**Apple Store:** the column **'prime_genre'** seems to show us what type the app is. 


In [24]:
def freq_table(dataset, index):
    frequency_table = {}
    total = 0
    for row in dataset:
        total += 1
        genre = str(row[index])
        if genre in frequency_table:
            frequency_table[genre] += 1
        else:
            frequency_table[genre] = 1
    percentage_table = {}
    for key in frequency_table:
        key_value = (frequency_table[key]/total)*100
        percentage_table[key] = key_value
    return percentage_table

In [25]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [26]:
prime_genre = display_table(dataset = appdata_freeapps, index = 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


As we can see above, for Apple Store, considering all the filters we applied, the most common genre is by far 'Games' (58.16%), following by Entertainment (7.88%). It is possible to see a huge difference between the 1st and 2nd, being greater than the half of apps of this list. As the general impression, most of the apps are designed more for entertainment (games, photo & video, social networking, sports, music) than practical purposes (education, shopping, utilities, productivity, lifestyle).

However, analyzing only this table, we cannot recommend an app profile, since we can't say that a large number of apps for a particular genre also imply that apps of that genre generally have a large number of users (which is our main goal).

In [27]:
Category = display_table(dataset=googledata_freeapps, index=1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The big picture in Google Play (considering all the filters) seems to be significantly different in comparison to Apple Store. 'Family' is the most common genre with (18.91%), following by 'Game' (9.72%) and 'Tools' (8.46%). Unlikely Google Play, in Apple Store we don't see the categories 'Family' and 'Tools', which are probably splited in other subsets. The same way, we don't see 'Enterteinment' in Google Play categories and it's the 2nd most common genre in Apple Store.

In contrast to Apple Store, Google Play seems to have more apps for practical purposes (Family, Tools, Business, Productivity, Finance). However, using only this data, we cannot confirm that only because the category describes 'Family', for instance, that's not about other things too. 
The data here seems to be a bit more distributed between the categories. We don't see here a single category with a huge percentage.

In [28]:
Genres = display_table(dataset=googledata_freeapps, index=9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

This data shows us a different approach: for each app, it can has more than one genre, which means that our categories here are splited in these genres. So the data here is more granular than in 'Category'.

So far, we can observe that Apple Store is more concentrated in enterteinment apps, while Google Play is more balanced between practical and fun apps. However, anyway these frequency tables don't reveal what genres have the most users, they only show us the frequency of app genres.

In [29]:
unique_genres = freq_table(appdata_freeapps, 11)

for genre in unique_genres:
    total = 0
    len_genre = 0
    for row in appdata_freeapps:
        genre_app = row[11]
        if genre_app == genre:
            nmb_user_rat = float(row[5])
            total += nmb_user_rat
            len_genre += 1
    avg_nmb_user_rat = total/len_genre
    print(str(genre) + ' = ' + str(avg_nmb_user_rat))
        

Social Networking = 71548.34905660378
Photo & Video = 28441.54375
Games = 22788.6696905016
Music = 57326.530303030304
Reference = 74942.11111111111
Health & Fitness = 23298.015384615384
Weather = 52279.892857142855
Utilities = 18684.456790123455
Travel = 28243.8
Shopping = 26919.690476190477
News = 21248.023255813954
Navigation = 86090.33333333333
Lifestyle = 16485.764705882353
Entertainment = 14029.830708661417
Food & Drink = 33333.92307692308
Sports = 23008.898550724636
Book = 39758.5
Finance = 31467.944444444445
Education = 7003.983050847458
Productivity = 21028.410714285714
Business = 7491.117647058823
Catalogs = 4004.0
Medical = 612.0


In order to see what genres have the most users, we can find this information in the 'Installs' column, but this information is missing for the App Store data set. Then, we'll take the total number of user ratings as a proxy in 'rating_count_column'.

If we blindly pick up the app with the higher average number of user rating, we would choose Navigation (avg = 86,090) as the most recommended genre. However, as we can see in the previous data, Navigation is one of the last places in frequency of app genres (0.19%).


In [30]:
for row in appdata_freeapps:
    appgenre = row[11]
    if appgenre == 'Navigation':
        print(str(row[1]) + ' = ' + str(row[5]))

Waze - GPS Navigation, Maps & Real-time Traffic = 345046
Google Maps - Navigation & Transit = 154911
Geocaching® = 12811
CoPilot GPS – Car Navigation & Offline Maps = 3582
ImmobilienScout24: Real Estate Search in Germany = 187
Railway Route Search = 5


As we can see above, the high average of this genre is heavily influenced by Waze (345046); Google Maps (154911) and the small number in denominator in average calculation: only 6 apps.

The same pattern we can see below for Social Networking apps (71,548): 

In [31]:
for row in appdata_freeapps:
    appgenre = row[11]
    if appgenre == 'Social Networking':
        print(str(row[1]) + ' = ' + str(row[5]))

Facebook = 2974676
Pinterest = 1061624
Skype for iPhone = 373519
Messenger = 351466
Tumblr = 334293
WhatsApp Messenger = 287589
Kik = 260965
ooVoo – Free Video Call, Text and Voice = 177501
TextNow - Unlimited Text + Calls = 164963
Viber Messenger – Text & Call = 164249
Followers - Social Analytics For Instagram = 112778
MeetMe - Chat and Meet New People = 97072
We Heart It - Fashion, wallpapers, quotes, tattoos = 90414
InsTrack for Instagram - Analytics Plus More = 85535
Tango - Free Video Call, Voice and Chat = 75412
LinkedIn = 71856
Match™ - #1 Dating App. = 60659
Skype for iPad = 60163
POF - Best Dating App for Conversations = 52642
Timehop = 49510
Find My Family, Friends & iPhone - Life360 Locator = 43877
Whisper - Share, Express, Meet = 39819
Hangouts = 36404
LINE PLAY - Your Avatar World = 34677
WeChat = 34584
Badoo - Meet New People, Chat, Socialize. = 34428
Followers + for Instagram - Follower Analytics = 28633
GroupMe = 28260
Marco Polo Video Walkie Talkie = 27662
Miitomo = 2

However in this case, we have a lot of apps, but the seven firsts dominates the statistic: Facebook, Pinterest, Skype for iPhone, Messenger, Tumblr, WhatsApp and Kik. Furthermore, Social Networking is a very red ocean to join in; very expensive to keep updating and hard to constantly inovating.

The same happens to Music (57,26), apps like Spotify, Shazam and Pandora also incorporates the high average for this genre.

Our goal is to find popular apps, but it seems that Music, Social Networking and Navigation are genres very hard to compete. The high average number of some genres seems to be heavily influenced by few apps with hundreds of thousands user ratings.

In [32]:
for row in appdata_freeapps:
    appgenre = row[11]
    if appgenre == 'Reference':
        print(str(row[1]) + ' = ' + str(row[5]))

Bible = 985920
Dictionary.com Dictionary & Thesaurus = 200047
Dictionary.com Dictionary & Thesaurus for iPad = 54175
Google Translate = 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran = 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition = 17588
Merriam-Webster Dictionary = 16849
Night Sky = 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) = 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools = 4693
GUNS MODS for Minecraft PC Edition - Mods Tools = 1497
Guides for Pokémon GO - Pokemon GO News and Cheats = 826
WWDC = 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free = 718
VPN Express = 14
Real Bike Traffic Rider Virtual Reality Glasses = 8
教えて!goo = 0
Jishokun-Japanese English Dictionary & Translator = 0


The average of Reference genre (74,942) above is also influenced by few apps (Bible and Dictionary). Even though, it can be a potential genre to be explored, since we could add some other popular book and turn it into an app, where we could add different features besides the raw version of the book. This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

- Weather apps: people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink: examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps: these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

In [33]:
googleplay_category = freq_table(googledata_freeapps, 1)
for category in googleplay_category:
    total = 0
    len_category = 0
    for row in googledata_freeapps:
        category_app = row[1]
        if category_app == category:
            installs = row[5]
            installs = installs.replace('+', '')
            installs = installs.replace(',', '')
            installs = float(installs)
            total += installs
            len_category += 1
    avg_nmb_installs = total/len_category
    print(category + ': ' + str(avg_nmb_installs))

ART_AND_DESIGN: 1986335.0877192982
AUTO_AND_VEHICLES: 647317.8170731707
BEAUTY: 513151.88679245283
BOOKS_AND_REFERENCE: 8767811.894736841
BUSINESS: 1712290.1474201474
COMICS: 817657.2727272727
COMMUNICATION: 38456119.167247385
DATING: 854028.8303030303
EDUCATION: 1833495.145631068
ENTERTAINMENT: 11640705.88235294
EVENTS: 253542.22222222222
FINANCE: 1387692.475609756
FOOD_AND_DRINK: 1924897.7363636363
HEALTH_AND_FITNESS: 4188821.9853479853
HOUSE_AND_HOME: 1331540.5616438356
LIBRARIES_AND_DEMO: 638503.734939759
LIFESTYLE: 1437816.2687861272
GAME: 15588015.603248259
FAMILY: 3695641.8198090694
MEDICAL: 120550.61980830671
SOCIAL: 23253652.127118643
SHOPPING: 7036877.311557789
PHOTOGRAPHY: 17840110.40229885
SPORTS: 3638640.1428571427
TRAVEL_AND_LOCAL: 13984077.710144928
TOOLS: 10801391.298666667
PERSONALIZATION: 5201482.6122448975
PRODUCTIVITY: 16787331.344927534
PARENTING: 542603.6206896552
WEATHER: 5074486.197183099
VIDEO_PLAYERS: 24727872.452830188
NEWS_AND_MAGAZINES: 9549178.467741935
MA

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

Again, if we blindly pick up a genre only for its number os installs, we would have the same problem as Apple Store:

In [34]:
for row in googledata_freeapps:
    category = row[1]
    if category == 'SOCIAL':
        print(str(row[0]) + ': ' + str(row[5]))
    

Facebook: 1,000,000,000+
Facebook Lite: 500,000,000+
Tumblr: 100,000,000+
Social network all in one 2018: 100,000+
Pinterest: 100,000,000+
TextNow - free text + calls: 10,000,000+
Google+: 1,000,000,000+
The Messenger App: 1,000,000+
Messenger Pro: 1,000,000+
Free Messages, Video, Chat,Text for Messenger Plus: 1,000,000+
Telegram X: 5,000,000+
The Video Messenger App: 100,000+
Jodel - The Hyperlocal App: 1,000,000+
Hide Something - Photo, Video: 5,000,000+
Love Sticker: 1,000,000+
Web Browser & Fast Explorer: 5,000,000+
LiveMe - Video chat, new friends, and make money: 10,000,000+
VidStatus app - Status Videos & Status Downloader: 5,000,000+
Love Images: 1,000,000+
Web Browser ( Fast & Secure Web Explorer): 500,000+
SPARK - Live random video chat & meet new people: 5,000,000+
Golden telegram: 50,000+
Facebook Local: 1,000,000+
Meet – Talk to Strangers Using Random Video Chat: 5,000,000+
MobilePatrol Public Safety App: 1,000,000+
💘 WhatsLov: Smileys of love, stickers and GIF: 1,000,000+

The average number of installs category "SOCIAL" (the third place with 23,253,652) seems to have been leveraged by few apps with more than 100,000,000+ installs.

In [35]:
for app in googledata_freeapps:
    if app[1] == 'SOCIAL' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Facebook : 1,000,000,000+
Facebook Lite : 500,000,000+
Tumblr : 100,000,000+
Pinterest : 100,000,000+
Google+ : 1,000,000,000+
Badoo - Free Chat & Dating App : 100,000,000+
Tango - Live Video Broadcast : 100,000,000+
Instagram : 1,000,000,000+
Snapchat : 500,000,000+
LinkedIn : 100,000,000+
Tik Tok - including musical.ly : 100,000,000+
BIGO LIVE - Live Stream : 100,000,000+
VK : 100,000,000+


We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for the genre which has the higher average number of installs: Communication (where we have giants like WhatsApp, Gmail, Skype, Google Chrome, etc.), and others like photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:

In [36]:
for app in googledata_freeapps:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [39]:
for app in googledata_freeapps:
    if (app[1] == 'BOOKS_AND_REFERENCE') and (app[5] == '100,000,000+' or
                                              app[5] == '500,000,000+' or
                                              app[5] == '1,000,000,000+'):
        print(app[0], ': ', app[5])

Google Play Books :  1,000,000,000+
Bible :  100,000,000+
Amazon Kindle :  100,000,000+
Wattpad 📖 Free Books :  100,000,000+
Audiobooks from Audible :  100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):


In [40]:
for app in googledata_freeapps:
    if (app[1] == 'BOOKS_AND_REFERENCE') and (app[5] == '1,000,000+' or
                                              app[5] == '5,000,000+' or
                                              app[5] == '10,000,000+' or
                                              app[5] == '50,000,000+'):
        print(app[0], ': ', app[5])

Wikipedia :  10,000,000+
Cool Reader :  10,000,000+
Book store :  1,000,000+
FBReader: Favorite Book Reader :  10,000,000+
Free Books - Spirit Fanfiction and Stories :  1,000,000+
AlReader -any text book reader :  5,000,000+
FamilySearch Tree :  1,000,000+
Cloud of Books :  1,000,000+
ReadEra – free ebook reader :  1,000,000+
Ebook Reader :  5,000,000+
Read books online :  5,000,000+
eBoox: book reader fb2 epub zip :  1,000,000+
All Maths Formulas :  1,000,000+
Ancestry :  5,000,000+
HTC Help :  10,000,000+
Moon+ Reader :  10,000,000+
English-Myanmar Dictionary :  1,000,000+
Golden Dictionary (EN-AR) :  1,000,000+
All Language Translator Free :  1,000,000+
Aldiko Book Reader :  10,000,000+
Dictionary - WordWeb :  5,000,000+
50000 Free eBooks & Free AudioBooks :  5,000,000+
Al-Quran (Free) :  10,000,000+
Al Quran Indonesia :  10,000,000+
Al'Quran Bahasa Indonesia :  10,000,000+
Al Quran Al karim :  1,000,000+
Al Quran : EAlim - Translations & MP3 Offline :  5,000,000+
Koran Read &MP3 30

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusions
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.